tests/instancetype/upgrade.go

```diff
@@ -0,0 +1,529 @@
+		Expect(err).ToNot(HaveOccurred())
+
+		vmi := libvmi.NewCirros()
+		removeResourcesAndPreferencesFromVMI(vmi)
+
+		vm = tests.NewRandomVirtualMachine(vmi, false)
+		vm.Spec.Instancetype = &virtv1.InstancetypeMatcher{
+			Name: instancetype.Name,
+			Kind: instancetypeapi.SingularResourceName,
+		}
+		vm.Spec.Preference = &virtv1.PreferenceMatcher{
+			Name: preference.Name,
+			Kind: instancetypeapi.SingularPreferenceResourceName,
+		}
+
+		vm, err = virtClient.VirtualMachine(testsuite.GetTestNamespace(instancetype)).Create(context.Background(), vm)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Wait for the initial revisionNames to be populated before we start out tests
+		libinstancetype.WaitForVMInstanceTypeRevisionNames(vm.Name, virtClient)
+	})
+
+	DescribeTable("ControllerRevisionUpgrade should upgrade", func(generateControllerRevision func() (*appsv1.ControllerRevision, error), updateMatcher func(string), getVMRevisionName func() string) {
+		// Capture the original RevisionName
+		originalRevisionName := getVMRevisionName()
+
+		By("Generating the target ControllerRevision")
+		cr, err := generateControllerRevision()
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Updating the VirtualMachine to reference the generated ControllerRevision")
+		OriginalTestRevisionName := cr.Name
+		updateMatcher(OriginalTestRevisionName)
+
+		By("Waiting for the ControllerRevision referenced by the VirtualMachine to be upgraded to the latest version")
+		var revisionName string
+		Eventually(func(g Gomega) {
+			By("Waiting for the RevisionName to be updated")
+			revisionName = getVMRevisionName()
+			g.Expect(revisionName).ToNot(Equal(originalRevisionName))
+
+			cr, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(context.Background(), revisionName, metav1.GetOptions{})
+			g.Expect(err).ToNot(HaveOccurred())
+
+			By("Ensuring the referenced ControllerRevision has the latest version label")
+			g.Expect(cr.Labels).To(HaveKeyWithValue(instancetypeapi.ControllerRevisionObjectVersionLabel, instancetypeapi.LatestVersion))
+
+			By("Ensuring the referenced ControllerRevision contains an object of the latest version")
+			decodedObj, err := runtime.Decode(generatedscheme.Codecs.UniversalDeserializer(), cr.Data.Raw)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(decodedObj.GetObjectKind().GroupVersionKind().Version).To(Equal(instancetypeapi.LatestVersion))
+
+		}, 30*time.Second, time.Second).Should(Succeed())
+
+		// If a new CR has been created assert that the old CR has been deleted
+		if OriginalTestRevisionName != revisionName {
+			_, err := virtClient.AppsV1().ControllerRevisions(vm.Namespace).Get(context.Background(), OriginalTestRevisionName, metav1.GetOptions{})
+			Expect(err).Should(HaveOccurred())
+			Expect(errors.ReasonForError(err)).Should(Equal(metav1.StatusReasonNotFound))
+		}
+
+	},
+		Entry("VirtualMachineInstancetype from v1beta1 without labels to latest",
```

Comment:
Should VirtualMachineInstanceTypeSpecRevisions be updated too?

Reply:
Yeah I'll pull in the functional tests for that as well now.